Issue #2840864: Implement the order summary totals service.

steveoliver opened the pull request:
See https://www.drupal.org/node/2840864.

steveoliver committed:
Issue #2840864: Implement the order summary totals service.

mglaman commented on modules/order/src/Entity/Order.php:
```diff
@@ -422,6 +445,43 @@ public function preSave(EntityStorageInterface $storage) {
   /**
+   * {@inheritdoc}
+   */
+  public function collectAdjustments() {
```
I know this is what @bojanz wanted, but it feels weird to have this here and then the next change where we don't just do `::getAdjustments()` directly. This feels like it belongs in a service to get the adjustments and their type metadata

mglaman replied:
Like it should just be part of OrderTotalSummary.

bojanz replied:
You need a way to get all order adjustments, both the ones on the order-level and the ones on the order-item level. getAdjustments() doesn't give you that.
This is a by-product of storing order-item-level adjustments on the order items themselves. Having it on the OrderTotalSummary would make sense if the Order class itself didn't need it.

mglaman replied:
Ah that's right. I forgot we need a way for the order to get all adjustments from its items. I guess I just got thrown off by the content of the method, thinking we could offload it.

mglaman commented on modules/order/src/Entity/Order.php:
```diff
@@ -440,13 +500,9 @@ protected function recalculateTotalPrice() {
     }
-    foreach ($this->getAdjustments() as $adjustment) {
-      $total_price = $total_price->add($adjustment->getAmount());
+    foreach ($this->collectAdjustments() as $adjustment) {
```
See above, feels weird.

mglaman commented on modules/order/src/OrderTotalSummary.php:
```diff
@@ -0,0 +1,87 @@
+    $total = $order->getTotalPrice();
+    $subtotal = $order->getSubtotalPrice();
+    $collected_adjustments = $order->collectAdjustments();
+    $types = $this->adjustmentTypeManager->getDefinitions();
```
collectAdjustments gets definitions, then we do again here. Furhter why I think collectAdjustments should be removed / moved here.

bojanz replied:
We can remove the definitions from collectAdjustment(), question was where to filter out adjustments with invalid types. If we want to make client classes responsible for it, that's fine.

mglaman replied:
ah, gotcha. yeah I suppose we do not want adjustments existing with invalid types :/

bojanz replied:
I guess it makes sense to move the type check to Adjustment::__construct and remove it from collectAdjustments(). That way we explode right away instead of silently ignoring broken data, probably a better idea.

steveoliver replied:
I did move the [validation of type into the Adjustment class](https://github.com/drupalcommerce/commerce/pull/594/files#diff-8361968a5f1463fc9c3e3043ba5b7c43) itself, with [a test for the expected exception](https://github.com/drupalcommerce/commerce/pull/594/files#diff-11a14da6d844f9a1efd7e467b1b41a93).

steveoliver replied:
We still need the adjustment type manager here though to get the weights for ordering.

mglaman commented on modules/order/src/OrderTotalSummary.php:
```diff
@@ -0,0 +1,87 @@
+          'items' => [],
+        ];
+      }
+      if (is_null($adjustment['source_id'])) {
```
why not just `empty()`

mglaman commented on modules/order/templates/commerce-order-total-summary.html.twig:
```diff
@@ -0,0 +1,31 @@
+  {{ 'Subtotal'|t }}: {{ totals.subtotal|commerce_price_format }}<br/>
+  {% for group in totals.adjustments %}
+    {% for item in group.items %}
+      {{ item.label }}: {{ item.amount|commerce_price_format }}<br/>
```
We should do some basic `div` splitting it up, or something inline. Spans? so it can be easily styled w/o overriding

bojanz replied:
Yeah, we need something that resembles the 1.x styling (two columns).

mglaman commented on modules/order/tests/src/Kernel/Entity/OrderTest.php:
```diff
@@ -105,6 +107,9 @@ protected function setUp() {
    */
   public function testOrder() {
+    /** @var \Drupal\commerce_order\OrderTotalSummaryInterface $order_total_summary */
+    $order_total_summary = \Drupal::service('commerce_order.order_total_summary');
```
Order summary should go in its own kernel?

mglaman reviewed (changes requested):
feels weird having the collect adjustment methods internals within the order class, willing to debate

steveoliver commented:
I addressed all the comments @mglaman and @bojanz made above.  The new kernel test has notes about its fails--even though the totals reflect the adjustments, adjustments with the custom 'test_adjustment_type' do not show up in the 'adjustments' of OrderTotalSummary::buildTotals; I've tried adding and using other types, only 'custom' ever shows up in the adjustments list... :/

steveoliver commented on modules/order/tests/src/Kernel/OrderTotalSummaryTest.php:
```diff
@@ -0,0 +1,176 @@
+      'unit_price' => new Price('12.00', 'USD'),
+    ]);
+    $order_item->save();
+
```
subtotal: 24.00
total: 24.00

steveoliver commented on modules/order/tests/src/Kernel/OrderTotalSummaryTest.php:
```diff
@@ -0,0 +1,176 @@
+      'amount' => new Price('-1.00', 'USD'),
+    ]));
+    $order_item->save();
+
```
subtotal: 24.00
total: 22.00

steveoliver commented on modules/order/tests/src/Kernel/OrderTotalSummaryTest.php:
```diff
@@ -0,0 +1,176 @@
+      'label' => '50 cent item fee',
+      'amount' => new Price('0.50', 'USD'),
+    ]));
+
```
subtotal: 24.00
total: 23.00

steveoliver commented on modules/order/tests/src/Kernel/OrderTotalSummaryTest.php:
```diff
@@ -0,0 +1,176 @@
+      'label' => '5 dollars off',
+      'amount' => new Price('-5.00', 'USD'),
+    ]));
+
```
subtotal: 24.00
total: 18.00

steveoliver commented on modules/order/tests/src/Kernel/OrderTotalSummaryTest.php:
```diff
@@ -0,0 +1,176 @@
+      'label' => 'Handling fee',
+      'amount' => new Price('10.00', 'USD'),
+    ]));
+
```
subtotal: 24.00
total: 28.00

steveoliver reviewed (commented):
Just pointing out why I think the amount assertions are correct.

bojanz commented:
This was committed in ea1ce2a2d57a959ea0817b2bf81ad0941a6ca79f
